nep-11.mediawiki

```diff
@@ -0,0 +1,256 @@
+
+The function SHOULD check whether the <code>from</code> address equals the caller contract hash.
+If so, the transfer SHOULD be processed; If not, the function SHOULD use the SYSCALL <code>Neo.Runtime.CheckWitness</code> to verify the transfer.
+
+If the <code>to</code> address is a deployed contract, the function SHOULD check the <code>payable</code> flag of this contract to decide whether it should transfer the tokens to this contract.
+
+If the transfer is not processed, the function SHOULD return <code>false</code>.
+
+====ownerOf====
+
+<pre>
+public static enumerator ownerOf(byte[] tokenid)
+</pre>
+
+Returns an <code>enumerator</code> that contains all the co-owners that own the specified token.
+
+The parameter <code>tokenid</code> SHOULD be a valid NFT.
+If not, this method SHOULD <code>throw</code> an exception.
+
+====balanceOf====
+
+<pre>
+public static BigInteger balanceOf(byte[] owner)
+</pre>
+
+Returns the total amount of NFTs owned by the specified address.
+
+The parameter <code>owner</code> SHOULD be a 20-byte address.
+If not, this method SHOULD <code>throw</code> an exception.
+
+<pre>
+public static BigInteger balanceOf(byte[] owner, byte[] tokenid)
+</pre>
+
+If the second parameter <code>tokenid</code> is also provided, this method returns the balance of the specified token for the specified owner's account.
+
+The parameter <code>tokenid</code> SHOULD be a valid NFT.
+If not, this method SHOULD <code>throw</code> an exception.
+
+The parameter <code>owner</code> SHOULD be a 20-byte address.
+If not, this method SHOULD <code>throw</code> an exception.
+
+If the <code>owner</code> is an unused address, or it's not the owner of the specified token, this method SHOULD return <code>0</code>.
+
+====tokensOf====
+
+<pre>
+public static enumerator tokensOf(byte[] owner)
+</pre>
+
+Returns an <code>enumerator</code> that contains all the tokens owned by the specified address.
+
+The parameter <code>owner</code> SHOULD be a 20-byte address.
+If not, this method SHOULD <code>throw</code> an exception.
+
+The returned tokens should be in the following format:
+<pre>
+{
+  "id": <ID>,
+  "uri": tokenURI(<ID>),
+  "properties": properties(<ID>),
+  ...
+}
+</pre>
+
+It must have the fields <code>id</code>, <code>uri</code> and <code>properties</code>.
+It can contain other dApp-specific fields.
+
+====tokenURI====
+
+<pre>
+public static string tokenURI(byte[] tokenid)
+</pre>
+
+Returns a distinct Uniform Resource Identifier (URI) for a given asset.
+URIs are defined in [https://tools.ietf.org/html/rfc3986 RFC 3986].
+This URI contains optionally mutable properties for the given NFT.
+
+The parameter <code>tokenid</code> SHOULD be a valid NFT.
+If not, this method SHOULD <code>throw</code> an exception.
+
+The URI MUST point to a JSON file that conforms to the "NEO NFT Metadata JSON Schema":
```

Comment:
Both `tokenURI()` and `properties()` return the "NEO NFT Metadata JSON Schema"?

Reply:
`tokenURI()` can return an http/https or data that points to a JSON following the schema. `properties()` returns an NVM object serialized in the same format.

Reply:
And they have the same fields? This makes me feel very strange.

Reply:
Why? Won't it make it easier to parse for the client if the two are in the same format?

Reply:
Both them have the same fields, which one is correct? For example, I return different name from `tokenURI()` and `properties()`. This is very confusing.

Reply:
Was a misunderstanding